Include only PackageReferences included at runtime in deps.json

Forgind opened the pull request:
Fixes https://github.com/dotnet/roslyn/issues/76797
Fixes #48944 

PrivateAssets isn't what we should be looking for to determine that we don't need an assembly in the deps.json file. This instead looks for ExcludeAssets and IncludeAssets to see if it will be in the output directory and includes it in the deps.json if so.

Forgind committed:
Include only PackageReferences included at runtime

Forgind committed:
Tweak test

ericstj commented:
I think this is the right change and better than the PrivateAssets approach.  I do want to mark this as "breaking" though because it's possible for folks to notice.  I would expect that in all the cases where folks notice it's really a no-op since I would expect all cases the SDK was just writing an empty entry.

The sort of things that folks *might* see - an entry goes away for a particular package, there's a package reference listed to a package but no definition.  For the latter perhaps the SDK can erase that reference too.

One thing that would be great to include with this is to also remove the entries that were dropped by conflict resolution, which is effectively another "automated" way that the SDK applies the equivalent of ExcludeAssets.

dsplaisted commented:
I think it might be better if instead of, or in addition to this, we did the following: Remove any Libraries from the deps.json that have no assets under them.  From the example in #39400:

![image](https://github.com/dotnet/sdk/assets/7412651/1c5ad178-8a4b-45c7-9d5b-5ee11cb18780)

NerdBank.GitVersioning is listed as a library under the .NET 7 target, but has no runtime assets or anything else under it.  So maybe we could remove it regardless of whether ExcludeAssets was specified, since whatever the cause it doesn't contribute anything at runtime.

This would require diving into the DependencyContextBuilder code, understanding how it works, and updating it.  We also might need to consider cases such as if a library has no assets but does list dependencies... do we need to check if it transitively brought in any assets?

Thoughts?

baronfel commented:
One thing I like about your addition @dsplaisted is that it makes the deps.json more purpose-built - the file _only_ cares about libraries and assets that impact the runtime experience.

For users that need accurate recounting of compile-time library dependencies we should be promoting SBOMs, which are a separate artifact specifically for book-keeping purposes.

dsplaisted commented:
@ericstj @ViktorHofer @agocke Any feedback on my [idea above](https://github.com/dotnet/sdk/pull/46218#issuecomment-2613346085) on removing libraries from the deps.json file if they don't have any assets?

What are the different asset types possible, and how should we handle them and handle dependencies between libraries that don't have any assets?

ericstj commented:
> @ericstj @ViktorHofer @agocke Any feedback on my [idea above](https://github.com/dotnet/sdk/pull/46218#issuecomment-2613346085) on removing libraries from the deps.json file if they don't have any assets?

Yes! this would be great.  We've wanted these gone for some time as they also cause false positives for folks scanning the files.

> What are the different asset types possible

The runtime cares about `runtime`, `native`, `runtimeTargets`, and `resources` @elinor-fung @jkoritzinsky 

> how should we handle them and handle dependencies between libraries that don't have any assets?

I think we just remove the library section and remove references to the removed section.  This should ensure that that an entire branch from the package graph is removed.  I can imagine if we're sloppy about identifying packages to remove it could lead to "orphaned" references - though I don't think the host would care about that since it doesn't attempt to construct a graph.

dsplaisted commented:
@ericstj How should we handle dependencies if we're removing packages without assets?  Should we remove a package that has no assets but depends on a package which does?  I think that could lead to a different kind of orphaning where we have a dependency of the removed package still in the file but no dependencies to it any more.

ericstj commented:
> Should we remove a package that has no assets but depends on a package which does?

I'd correct this to say 
> package that has no assets but is the only package which depends on another package that has assets
This is not so uncommon if you consider a case where someone does ExcludeAssets on one package with a large package graph, but then updates one of that package's dependencies to a newer version with a direct reference (or some other graph) that's not excluded.

If you could constrain the removal in this way efficiently then I think it would be safer, however my hypothesis is that this is unlikely if you are identifying removal based on exclusion/conflict resolution.  Package exclusion is specified by graph notation which flows to dependencies and conflict resolution is closure complete.  So you could count on these mechanisms being graph complete for a given path.

I can see how if instead we remove packages that have no assets (regardless of how) then it would be safer to also check that they are not the only packages that reference some other package with assets.

dsplaisted commented:
@Forgind Are you thinking of taking a look at this again?  I think what we should try to do what I suggested [here](https://github.com/dotnet/sdk/pull/46218#issuecomment-2613346085), ie remove libraries from the file that don't have any assets associated with them.

Forgind commented:
I talked with dsplaisted offline, and I think we're on the same page. I'm planning to start working on this again early next week.

Forgind committed:
First step in fixing deps.json

Forgind commented:
@dsplaisted, can you look at that last commit (https://github.com/dotnet/sdk/pull/46218/commits/fa1bd9d3a5fd8ed9a3df3a3476fae877d4ae9c01) to see if you think that aligns with your understanding of what I should do, including the comment? If so, I'll turn the comment into code.

dsplaisted commented on src/Tasks/Microsoft.NET.Build.Tasks/DependencyContextBuilder.cs:
```diff
@@ -292,23 +292,34 @@ public DependencyContext Build(string[] userRuntimeAssemblies = null)
                 runtimeLibraries.Add(runtimeLibrary);
             }
 
+            /*
+             * We now need to modify runtimeLibraries to eliminate those that don't have any runtime assets. We follow the following steps:
+             * 0. Construct a reverse dependencies list: all runtimeLibraries that depend on this one
+             * 1. If runtimeAssemblyGroups, nativeLibraryGroups, dependencies, and resourceAssemblies are all empty, remove this runtimeLibrary as well as any dependencies on it.
+             * 2. Add all runtimeLibraries to a list of to-be-processed libraries called libraryCandidatesForRemoval
+             * 3. libraryCandidatesForRemoval.Pop() --> if there are no runtimeAssemblyGroups, nativeLibraryGroups, or resourceAssemblies, and either dependencies is empty or all
+             *      dependencies have something else that depends on them, remove it (and from libraryCandidatesForRemoval), adding everything that depends on this to 
+             *      libraryCandidatesForRemoval if it isn't already there
+             * Repeat 3 until libraryCandidatesForRemoval is empty
+             */
+
```
Yes, this looks like it should work.

It may make sense to update ModifiableRuntimeLibrary to have lists of both the dependencies and the dependants (the other direction) so that you can use that as the graph of objects to walk.

Instead of a stack you may also want to use something like a HashSet so you can put a candidate for removal in the list for processing without having to process it multiple times if more than one library that gets removed has it as a dependant before you get to processing that library.

Forgind committed:
Turn comment into code

Forgind commented:
@dsplaisted, I think I correctly turned the comment into code. This PR still includes the include/exclude logic. I think that's correct, so I'd lean towards keeping it; do you have any problem with leaving it in?

If not, I'll just resolve the conflicts, and this should be good to review

Forgind committed:
Merge branch 'release/9.0.3xx' of https://github.com/dotnet/sdk into deps-json-has-extant-assemblies

Forgind committed:
remove unnecessary change

Forgind committed:
Guard against missing deps

dsplaisted commented:
I haven't looked in detail at the graph algorithm code yet, but we are going to want some decent test coverage for this.  We can probably have an end-to-end test or two (#39400 has a good example), and then unit tests where we test just the task with hand-modified assets files for different interesting cases.

Forgind committed:
Add test

Forgind committed:
Remove ExcludeAssets

Forgind committed:
Add mini unit test

Forgind committed:
Add more complicated unit test

This one prunes out 3/7 references

Forgind commented:
> I haven't looked in detail at the graph algorithm code yet, but we are going to want some decent test coverage for this. We can probably have an end-to-end test or two (#39400 has a good example), and then unit tests where we test just the task with hand-modified assets files for different interesting cases.

I added some tests. Let me know what you think!

dsplaisted commented on src/Tasks/Microsoft.NET.Build.Tasks/DependencyContextBuilder.cs:
```diff
@@ -885,5 +944,17 @@ private struct LibraryDependency
+        private class ModifiableRuntimeLibrary
+        {
+            public RuntimeLibrary Library { get; set; }
+            public HashSet<string> Dependents { get; set; }
```
I think it's easy to get confused about what "dependent" means (libraries that depend on this one, ie the opposite of a dependency), so it would be good ta have a clarifying comment here.

dsplaisted commented on src/Tasks/Microsoft.NET.Build.Tasks/DependencyContextBuilder.cs:
```diff
@@ -285,23 +285,83 @@ public DependencyContext Build(string[] userRuntimeAssemblies = null)
+             *      libraryCandidatesForRemoval if it isn't already there
+             * Repeat 3 until libraryCandidatesForRemoval is empty
+             */
+            var references = runtimeLibraries.ToDictionary(lib => lib.Library.Name, lib => lib);
+            foreach (var reference in runtimeLibraries)
```
"Reference" seems to mean something slightly different than "library".  Is there a reason not to name these `libraries` and `library` instead of `references` and `reference`?

dsplaisted commented on src/Tasks/Microsoft.NET.Build.Tasks/DependencyContextBuilder.cs:
```diff
@@ -285,23 +285,83 @@ public DependencyContext Build(string[] userRuntimeAssemblies = null)
+                }
+            }
+
+            var unprocessedReferences = runtimeLibraries.ToHashSet();
+            HashSet<ModifiableRuntimeLibrary> temp = new();
```
What's the reason for the separate `temp` HashSet?  Originally I thought it was because you couldn't add to the `unprocessedReference` set while iterating over it, but now I see that you're not actually iterating over it, just popping an item out each time through the loop.  Given that, could you remove the separate `temp` list?

If not, it would probably be a good idea to rename `temp` to something more descriptive and add a comment about why it's used.

Forgind replied:
Good point. That's exactly how I'd been thinking about it, but you're right that it wasn't actually being iterated over, so there's no need for a temp HashSet. Deleted it

dsplaisted commented on src/Tasks/Microsoft.NET.Build.Tasks/DependencyContextBuilder.cs:
```diff
@@ -399,11 +458,11 @@ public DependencyContext Build(string[] userRuntimeAssemblies = null)
             _compilationOptions ?? CompilationOptions.Default,
             compilationLibraries,
-            runtimeLibraries,
+            runtimeLibraries.Select(library => library.Library),
```
Since the library and its dependencies are actually immutable, I think you need to recreate each library after the graph walk so that any dependencies that were removed don't show up as dangling links in the deps file.

dsplaisted reviewed (commented):
Great job with the graph algorithm code.  I went through it and I'm pretty sure it's doing what we discussed.

I haven't yet looked at the tests in detail, but I suspect we will need more of them.  We can start with the different cases that were mentioned in the discussions for this PR and linked issues, creating graphs that match those cases and then possibly thinking of other interesting variations.

Forgind committed:
PR feedback

dsplaisted commented:
@forgind Here are some test cases that I think we should cover in the `GivenADependencyContextBuilder` unit tests:

- Leaf nodes
  - Direct reference to package with no assets: Project -> HasNoAssets
  - Indirect reference to package with no assets: Project -> HasAssets -> HasNoAssets
- Dependency chains
  - Package with no assets references another package with no assets: Project -> HasNoAssets1 -> HasNoAssets2
  - Package with no assets references another package with assets: Project -> HasNoAssets -> HasAssets (they should all show up in the deps file)
- Dependency graphs
  - Package with no assets references a package with assets which is also referenced by project:
    - Project -> HasNoAssets -> HasAssets
    - Project -> HasAssets
    - HasNoAssets should not appear in deps file
  - Package with no assets reference a package referenced by another package
    - Project -> HasNoAssets -> HasAssets1
    - Project -> HasAssets2 -> HasAssets1
    - HasNoAssets should not appear in deps file
  - Two packages with no assets reference a package with assets:
    - Project -> HasNoAssets1 -> HasAssets
    - Project -> HasNoAssets2 -> HasAssets
    - This is an interesting case.  I think the behavior as implemented in the PR would be that exactly one of HasNoAssets1 and HasNoAssets2 would appear in the deps file.

There may also be some interesting test cases with packages that have compile-time assets but no runtime assets.

You will need an assets file for each of these test cases (it's called a lock file in these tests because they're really old).  You could author them manually, but there are enough different cases and we may want to add more that I would probably create code where you could pass it a graph and it would create the assets file for you.

Forgind committed:
Properly handle main project references

Forgind committed:
Add unit tests

Forgind committed:
Little tweaks (one per dsplaisted's suggestion)

Forgind committed:
e2e test

Forgind committed:
Make tests pass

For ItDoesntCreateReferenceAssembliesWhenNoCompilationOptions, the issue is that the RuntimeLibraries are minimized to remove libraries with no assets and no (transitive) dependencies with assets. This removes all but one of the RuntimeLibraries, the one remaining being the main project. Per dsplaisted's suggestion, I then removed "Dependencies" that are absent from Libraries, which means removing all of its dependencies, so RuntimeLibraries should now have only one entry, and it should have no dependencies, hence the shift.

dsplaisted commented on src/Tasks/Microsoft.NET.Build.Tasks.UnitTests/LockFiles/simpleApp.project.lock.json:
```diff
@@ -0,0 +1,6644 @@
+{
+    "locked": false,
```
This file is more than 6000 lines long. It looks like in the tests you're loading it and then adding the test libraries you need, so can you instead use an assets file that's as simple as possible from a project with no package references as the starting point?

Forgind replied:
Yep, pretty much. I'm not really sure if there's a better way to do this. I found other tests that were using a different project.lock.json (and even wrote a couple tests against it), but it had some extra dependencies that I didn't want that would've just been confusing. (I think they also made it impossible to properly test a couple scenarios like "no references.") I basically just copied what was there then deleted the parts I didn't want.

dsplaisted commented on src/Tasks/Microsoft.NET.Build.Tasks.UnitTests/GivenADependencyContextBuilder.cs:
```diff
@@ -211,6 +213,168 @@ public void ItHandlesReferenceAndPackageReferenceNameCollisions()
+        [Fact]
+        public void DirectReferenceToPackageWithNoAssets()
+        {
+            DependencyContext dependencyContext = BuildDependencyContextFromDependenciesWithResources([], [], ["System.A"]);
```
I'm not sure about how the references parameter is handled here... it looks like that's creating direct DLL references, and the tests may not model dependencies from the main project to a NuGet package, which is the more common scenario.

Forgind replied:
Yes; it's from the main project to any of its direct references. I'm not sure I understand the distinction you're making between a direct reference from the main project to a dll vs. a direct reference to a nuget package? Are you saying that with the nuget packages we'd be downloading it just before using it? I don't see how that distinction is represented in DependencyContextBuilder, though I may be missing something.

dsplaisted commented on src/Tasks/Microsoft.NET.Build.Tasks.UnitTests/GivenADependencyContextBuilder.cs:
```diff
@@ -211,6 +213,168 @@ public void ItHandlesReferenceAndPackageReferenceNameCollisions()
+                        Dependencies = kvp.Value.Select(n => new PackageDependency(n)).ToList()
+                    });
+
+                if (withResources.Contains(kvp.Key))
```
I think having resource assets is not the most common case, so instead we should probably model whether the libraries have runtime or compilation assets (and then resource assets if we have additional scenarios for that).

Forgind replied:
All three are handled in the same way in this part of the code, but they're added in different ways...I don't disagree with you, exactly, but I suspect it would take me a couple days to get that to work, and it wouldn't actually test anything we aren't already properly testing using resource assets.

dsplaisted commented on src/Tasks/Microsoft.NET.Build.Tasks/DependencyContextBuilder.cs:
```diff
@@ -399,11 +481,25 @@ public DependencyContext Build(string[] userRuntimeAssemblies = null)
+                library.Library.NativeLibraryGroups,
+                library.Library.ResourceAssemblies,
+                library.Library.Dependencies.Where(
+                    dependency => runtimeLibraries.Any(lib => lib.Library.Name.Equals(dependency.Name)) ||
+                    compilationLibraries.Any(lib => lib.Name.Equals(dependency.Name))).ToList(),
```
This looks like it ends up being O(N^2) or similar.  Since I think it may not be very rare to have projects with hundreds or thousands of dependencies, we should probably use a lookup (hash or dictionary) to check if a dependency should be included rather than doing a list scan for each one.

dsplaisted commented on test/Microsoft.NET.Build.Tests/GivenThatWeWantToBuildADesktopLibrary.cs:
```diff
@@ -105,10 +105,58 @@ public void Method1()
+            var buildCommand = new BuildCommand(testAsset);
+            buildCommand.Execute().Should().Pass();
+
+            using (var depsJsonFileStream = File.OpenRead(Path.Combine(buildCommand.GetOutputDirectory("net9.0").FullName, "PackageWithoutAssets_ShouldNotShowUpInDepsJson.deps.json")))
```
Avoid hardcoding the target framework in one place but not another place so this doesn't break for .NET 10 (or higher).

```suggestion
            using (var depsJsonFileStream = File.OpenRead(Path.Combine(buildCommand.GetOutputDirectory(ToolsetInfo.CurrentTargetFramework).FullName, "PackageWithoutAssets_ShouldNotShowUpInDepsJson.deps.json")))
```

dsplaisted reviewed (commented):
Good progress.

I think I'll want to go over the test cases added to GivenADependencyContextBuilder with you and make sure I understand them and that we are covering the right scenarios.

Forgind committed:
Merge branch 'main' of https://github.com/dotnet/sdk into deps-json-has-extant-assemblies

Forgind commented on src/Tasks/Microsoft.NET.Build.Tasks/DependencyContextBuilder.cs:
```diff
@@ -285,23 +285,106 @@ public DependencyContext Build(string[] userRuntimeAssemblies = null)
+                }
+            }
+
+            // Rather than adding the main project's dependencies, we added references to them, i.e., Foo.Reference.dll
+            // instead of Foo.dll. This adds Foo.dll as a reference directly so another reference can be removed if it
```
It's been a month since I wrote this, and it's a little confusing to me now. What I _think_ was happening and I was trying to explain in this comment is that it sometimes comes out as Foo.Reference.dll and sometimes comes out as Foo.dll, exactly as you said, but we don't really care which because in either case, there's a reference from the main project to Foo, and we want to make sure to track that.

The issue is that we might have missed that there was a reference to Foo if it did come out as Foo.Reference.dll earlier, in which case there's an edge missing in our graph, and this adds the edge. If it was already added, we're adding it a second time and relying on HashSet logic to eliminate the duplicate.

Forgind commented on src/Tasks/Microsoft.NET.Build.Tasks/DependencyContextBuilder.cs:
```diff
@@ -285,23 +285,106 @@ public DependencyContext Build(string[] userRuntimeAssemblies = null)
+
+            // Rather than adding the main project's dependencies, we added references to them, i.e., Foo.Reference.dll
+            // instead of Foo.dll. This adds Foo.dll as a reference directly so another reference can be removed if it
+            // isn't necessary because of a direct reference from the main project.
```
There were definitely failing test cases. I think this fixed one or two of them where there were .Reference.dlls that weren't being seen as dependencies.

Forgind replied:
The specific failing test was PackageWithNoAssetsReferencesPackageReferencesByOtherPackage. It has a structure like this:
main app -> A
main app -> B
A -> B
B has assets; A does not

In this case, since A does not have assets, and main app references B directly, it should be able to skip A and remove it from the deps.json, but without this piece, it doesn't.

Forgind commented on src/Tasks/Microsoft.NET.Build.Tasks/DependencyContextBuilder.cs:
```diff
@@ -285,23 +285,106 @@ public DependencyContext Build(string[] userRuntimeAssemblies = null)
+                    {
+                        if (libraries.TryGetValue(directReference.Name, out var dep))
+                        {
+                            dep.Dependents.Add(_mainProjectInfo.Name);
```
It's the opposite, actually; it tries to add Foo.Reference.dll when it should be adding Foo.dll. It stops short of actually adding it, though, because it looks in runtimeLibraries and doesn't find it, so it just shrugs and moves on.

I don't think it really matters to me whether we fix the dependents list in a second loop or just get it right in the first place, but it wasn't too hard to do, so I moved that logic up so we don't have the mistake in the first place.

Forgind committed:
PR comments

Forgind committed:
Support package references and cut down lock file

ilonatommy commented:
@Forgind it looks like this broke aspnetcore tests that use xunit and run with `dotnet test` command. This is an important bit for us, because currently all the tests on CI that use `dotnet test` discover 0 tests to run and finish with success.

We can see that there are 3 packages missing in `deps.json` after this change flew in:
- "xunit.analyzers" 
- "xunit.runner.console"
- "xunit.runner.visualstudio"

To repro:
- take aspnetcore repo, build and find any `.Test.csproj`, e.g.  `src\Http\WebUtilities\test\Microsoft.AspNetCore.WebUtilities.Tests.csproj`
- run `dotnet test src\Http\WebUtilities\test\Microsoft.AspNetCore.WebUtilities.Tests.csproj` that finishes with
```
warning : No test is available in C:\Users\itomkowicz\Desktop\aspnetcore-fork3\artifacts\bin\Microsoft.AspNetCore.WebUtilities.Tests\Debug\net10.0\Microsoft.AspNetCore.WebUtilities.Tests.dll. Make sure that test discoverer & executors are registered and platform & framework version settings are appropriate and try again.

Test summary: total: 0; failed: 0; succeeded: 0; skipped: 0; duration: 1,2s
Build succeeded with 1 warning(s) in 26,1s
```
- observe `artifacts\bin\Microsoft.AspNetCore.WebUtilities.Tests\Debug\net10.0\Microsoft.AspNetCore.WebUtilities.Tests.deps.json` that does not have
```
          "XunitXml.TestLogger": "2.1.26",
          "xunit.analyzers": "1.15.0",
          "xunit.assert": "2.9.2",
          "xunit.core": "2.9.2",
          "xunit.runner.console": "2.9.2",
          "xunit.runner.visualstudio": "2.8.2"
```

After reverting this change and re-testing `Microsoft.AspNetCore.WebUtilities.Tests`, we finish with
```
[xUnit.net 00:00:00.00] xUnit.net VSTest Adapter v2.8.2+699d445a1a (64-bit .NET 10.0.0-preview.6.25302.103)
[xUnit.net 00:00:00.15]   Discovering: Microsoft.AspNetCore.WebUtilities.Tests
[xUnit.net 00:00:00.35]   Discovered:  Microsoft.AspNetCore.WebUtilities.Tests
[xUnit.net 00:00:00.36]   Starting:    Microsoft.AspNetCore.WebUtilities.Tests
[xUnit.net 00:00:01.02]     Microsoft.AspNetCore.WebUtilities.FileBufferingReadStreamTests.Read_BufferingContentToDisk_CreatesFileWithUserOnlyUnixFileMode [SKIP]
[xUnit.net 00:00:01.02]       UnixFileMode is not supported on Windows.
[xUnit.net 00:00:01.08]     Microsoft.AspNetCore.WebUtilities.FileBufferingWriteStreamTests.Write_BufferingContentToDisk_CreatesFileWithUserOnlyUnixFileMode [SKIP]
[xUnit.net 00:00:01.08]       UnixFileMode is not supported on Windows.
[xUnit.net 00:00:01.09]   Finished:    Microsoft.AspNetCore.WebUtilities.Tests
  Microsoft.AspNetCore.WebUtilities.Tests test succeeded (2,8s)

Test summary: total: 478; failed: 0; succeeded: 476; skipped: 2; duration: 2,7s
Build succeeded in 24,9s
```

ilonatommy commented:
cc @dsplaisted
